src/Microsoft.AspNet.Server.Kestrel/Http/MessageBody.cs

```diff
@@ -38,35 +39,40 @@ public Task<int> ReadAsync(ArraySegment<byte> buffer, CancellationToken cancella
             return result;
         }
 
-        public async Task Consume(CancellationToken cancellationToken = default(CancellationToken))
+        public Task Consume(CancellationToken cancellationToken = default(CancellationToken))
         {
-            Task<int> result;
-            var send100checked = false;
-            do
+            var result = ReadAsyncImplementation(default(ArraySegment<byte>), cancellationToken);
+            if (!result.IsCompleted)
             {
-                result = ReadAsyncImplementation(default(ArraySegment<byte>), cancellationToken);
-                if (!result.IsCompleted)
-                {
-                    if (!send100checked)
-                    {
-                        if (Interlocked.Exchange(ref _send100Continue, 0) == 1)
-                        {
-                            _context.FrameControl.ProduceContinue();
-                        }
-                        send100checked = true;
-                    }
-                }
-                else if (result.GetAwaiter().GetResult() == 0) 
-                {
-                    // Completed Task, end of stream
-                    return;
-                }
-                else
+                if (Interlocked.Exchange(ref _send100Continue, 0) == 1)
                 {
-                    // Completed Task, get next Task rather than await
-                    continue;
+                    _context.FrameControl.ProduceContinue();
                 }
-            } while (await result != 0);
+
+                return ConsumeAwaited(result, cancellationToken);
+            }
+            else if (result.GetAwaiter().GetResult() == 0)
+            {
+                // Completed Task, end of stream
+                return TaskUtilities.CompletedTask;
+            }
+            else
+            {
+                // Completed Task, but non-zero get next Task and await
+                return ConsumeAwaited(ReadAsyncImplementation(default(ArraySegment<byte>), cancellationToken), cancellationToken);
```

Comment:
Yeah is a bit weird; enough to make me look at it for a good few minutes before working out why.

Is because the non-completed one passes in the current task so is to match that. Will have a think on how it can be made better.